File: server/fastapi/config/lifecycles/29_app_yaml_fetch_config.lifecycle.py
"""Bind FetchConfigHandle to app.state — terminal lifecycle in the chain."""

from __future__ import annotations

import logging

from app_yaml_fetch_config import EndpointConfigSDK, list_endpoints, load_config

from ._app_yaml_fetch_config import FetchConfigHandle

log = logging.getLogger("fastapi_server.app_yaml_fetch_config")


def on_init(app, _config) -> None:
    app.state.app_yaml_fetch_config = None


async def on_startup(app, _config) -> None:
    cfg_handle = getattr(app.state, "app_yaml_config", None)
    applier = getattr(app.state, "app_yaml_applier", None)
    if cfg_handle is None or applier is None:
        raise RuntimeError(
            "app_yaml_config or app_yaml_applier missing — confirm 27 + 28 run before 29"
        )
    merged = await applier(cfg_handle.raw)
    load_config(merged)
    sdk = EndpointConfigSDK()
    app.state.app_yaml_fetch_config = FetchConfigHandle(sdk=sdk)
    log.info("app_yaml_fetch_config initialized (endpoints=%d)", len(list_endpoints()))


async def on_shutdown(app, _config) -> None:
    app.state.app_yaml_fetch_config = None

File: server/fastify/config/lifecycles/29_app_yaml_fetch_config.lifecycle.mjs
import fp from "fastify-plugin";
import {
  EndpointConfigSDK,
  getEndpoint,
  getFetchConfig,
  listEndpoints,
  loadConfig,
  resolveIntent,
} from "@ployglot/app-yaml-fetch-config";

async function fetchConfigPlugin(fastify, _opts) {
  if (!fastify.app_yaml_config) {
    throw new Error(
      "app_yaml_config missing — confirm 27_app_yaml_config.lifecycle.mjs runs before 29"
    );
  }
  if (!fastify.app_yaml_applier) {
    throw new Error(
      "app_yaml_applier missing — confirm 28_app_yaml_from_context.lifecycle.mjs runs before 29"
    );
  }
  const merged = await fastify.app_yaml_applier(fastify.app_yaml_config.getAll());
  loadConfig(merged);

  // SDK gets the endpoint.dev.yaml path so refreshConfig() can re-read on demand
  // (slot 36 routes call sdk.refreshConfig()). Path source: APP_YAML_FIXTURES_DIR.
  const fixturesDir = process.env.APP_YAML_FIXTURES_DIR;
  const endpointFile = fixturesDir ? `${fixturesDir}/endpoint.dev.yaml` : null;
  const handle = {
    get_fetch_config: (intent, payload) => getFetchConfig(intent, payload ?? {}),
    list_endpoints: () => listEndpoints(),
    get_endpoint: (name) => getEndpoint(name),
    resolve_intent: (intent) => resolveIntent(intent),
    sdk: new EndpointConfigSDK({ filePath: endpointFile }),
  };

  fastify.decorate("app_yaml_fetch_config", handle);
  fastify.decorateRequest("app_yaml_fetch_config", null);
  fastify.addHook("onRequest", async (req) => {
    req.app_yaml_fetch_config = handle;
  });
}

const wrapped = fp(fetchConfigPlugin, {
  name: "app-yaml-fetch-config",
  fastify: ">=4",
  dependencies: ["app-yaml-config", "app-yaml-from-context"],
});

export default async function lifecycle(server, _config) {
  await server.register(wrapped);
  server.log.info("app-yaml-fetch-config decorator registered");
}
